src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Cache.cs

```diff
@@ -148,10 +148,10 @@ internal sealed class CachedCodeEntry
 #endif
             public int _capsize;
             public ExclusiveReference _runnerref;
-            public SharedReference _replref;
+            public WeakReference<RegexReplacement> _replref;
```

Comment:
Why is it a weak reference?  It it really big or something and thus we only want to cache it if it's being used?

Reply:
The easy answer: because that's how it was ported to Core.
RegexReplacement is not big at all. It stores the string pattern, a list of rules (strings) and a list of ints. It doesn't know anything about the input string, it just stores the different rules based on the parsed pattern. Usually very small inputs. I think your point is legit, we don't need a weak reference here.

Reply:
But it is put into the Regex Cache, it's part of the CacheCodeEntry. If it's not a WeakReference the cached will hold on it as long as the item exists. I think that was the original intent.

Reply:
I think a strong reference to RegexReplacement in Regex makes sense but the CachedCodeEntry that can live far longer than the Regex instance should have a weak one to avoid GC holes for very large cache sizes.

Reply:
I'm curious why we want to cache Regex objects (to save re-parsing) but not cache RegexReplacement objects (to save re-parsing) especially if they are not large. 

It's reasonable to say "because that's how we always did it" but it's not obvious to me why.

Reply:
Our engine is weird in some places...

Note that we don't store Regex objects but only some parts of it:

```csharp
public CachedCodeEntryKey _key; <-- obviously necessary to identify the entry
public RegexCode _code; <-- the entire generated code. takes time co compute
public Hashtable _caps; <-- precomputed same as code
public Hashtable _capnames;  <-- precomputed same as code
public string[] _capslist;  <-- precomputed same as code
#if FEATURE_COMPILED
public RegexRunnerFactory _factory; <-- I guess this is cached because this is comes from the user and could take very long to compute
#endif
public int _capsize; <-- easy to cache
public ExclusiveReference _runnerref; <-- to guarantee thread safety
public WeakReference<RegexReplacement> _replref;
```

Reply:
Even more obscure why the WeakReference was chosen here as the CachedCodeEntry is already very large and contains many references.